vllm/model_executor/models/utils.py

```diff
@@ -282,6 +282,24 @@ def is_pp_missing_parameter(name: str, model: torch.nn.Module) -> bool:
     return False
 
 
+def get_inputs_embeds(
+    input_ids: torch.Tensor,
+    embeddings_module: torch.nn.Module,
+    inputs_embeds: Optional[torch.Tensor] = None,
+    inputs_embeds_masks: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    """Get the input embeddings from either `input_ids` and `inputs_embeds`."""
+    if inputs_embeds is not None:
+        if inputs_embeds_masks.all().item():
+            hidden_states = inputs_embeds
+        else:
+            hidden_states = embeddings_module(input_ids)
+            hidden_states[inputs_embeds_masks] = inputs_embeds
+    else:
+        hidden_states = embeddings_module(input_ids)
+    return hidden_states
```

Comment:
I've looked at the code more and it looks like `inputs_embeds_masks` can only be all zeros or all ones? In that case it would be more performant to use a simple boolean flag.

Reply:
Any thoughts about this @comaniac?

Reply:
Doing so should also let this feature work with chunked prefill, similar to #8346.

Reply:
`inputs_embeds_masks` is a 1d tensor of booleans, and each value is whether the request has `input_embeds` or not

Reply:
From my understanding, the current `inputs_embeds_masks` has a size equal to the number of tokens. My suggestion is to change that to be the number of sequences.

Reply:
yes you are right. Currently the length of `inputs_embeds_masks` is equal to number of tokens. If we change it to number of sequences, we will need to iterate over `positions` to set the embeds like below
```
def get_inputs_embeds(
    input_ids: torch.Tensor,
    positions: torch.Tensor,
    embeddings_module: torch.nn.Module,
    inputs_embeds: Optional[torch.Tensor] = None,
    inputs_embeds_masks: Optional[torch.Tensor] = None,
) -> torch.Tensor:
    """Get the input embeddings from either `input_ids` and `inputs_embeds`."""
    if inputs_embeds is not None:
        if inputs_embeds_masks.all().item():
            hidden_states = inputs_embeds
        else:
            hidden_states = embeddings_module(input_ids)
            starting_positions = [idx for idx, v in enumerate(positions) if v == 0] + [len(positions)]
            embeds_position = 0
            for idx, has_embeds in enumerate(inputs_embeds_masks):
                if has_embeds:
                    seq_len = starting_positions[idx + 1] - starting_positions[idx]
                    hidden_states[starting_positions[idx]:starting_positions[idx + 1]] = inputs_embeds[embeds_position:embeds_position + seq_len]
                    embeds_position += seq_len
    else:
        hidden_states = embeddings_module(input_ids)
    return  hidden_states
```

Reply:
Does this still hold when some of the items in the mask are False?

Also, can you test both implementations with chunked prefill and see if they work?

Reply:
This error looks like the one @OswaldoBornemann got. @OswaldoBornemann can you turn off chunked prefill and see if it works?

We can leave chunked prefill to another PR. I'll test this PR locally when I have time and report back to you.

Reply:
The logic of `input_embeds` is quite complicated when multi-modal models are involved, let me see how to simplify this.

Reply:
Since we already have image embeddings for multi-modal models, I've disabled custom `input_embeds` for them to avoid conflicts between them.

Reply:
> We can leave chunked prefill to another PR (assuming that you didn't break the function of chunked prefill when input embeddings aren't being inputted). I'll test this PR locally when I have time and report back to you.

yes chunked prefill works when input_embeds is not used

Reply:
Yeah, I fixed it by adding an extra `inputs_embeds_masks is not None` check inside `get_inputs_embeds`. However, I found another problem when both `inputs_embeds` and image embeddings are passed together (they conflict with each other). Usually for multi-modal models, we only pass the image embeddings, so it is simpler to just restrict the inputs for now.

Reply:
I'm thinking of refactoring the inputs logic to be more reusable in a future PR, so we can add support then.